ember-core/src/main/scala/org/http4s/ember/core/Parser.scala

```diff
@@ -27,57 +27,41 @@ import scala.collection.mutable
 
 private[ember] object Parser {
 
-  object HeaderP {
+  final case class MessageP(bytes: Array[Byte], rest: Array[Byte])
 
-    def parseHeaders[F[_]](
-        head: Array[Byte],
-        read: F[Option[Chunk[Byte]]],
-        maxHeaderLength: Int,
-        acc: Option[ParseHeadersIncomplete])(implicit
-        F: MonadThrow[F]): F[(Headers, Boolean, Option[Long], Array[Byte])] = {
-      // TODO: improve this
-      val nextChunk = if (head.nonEmpty) F.pure(Some(Chunk.bytes(head))) else read
-      nextChunk.flatMap {
-        case Some(chunk) =>
-          val nextArr: Array[Byte] = acc match {
-            case None => chunk.toArray
-            case Some(last) => combineArrays(last.bv, chunk.toArray)
-          }
-          val result = acc match {
-            case None => headersInSection(nextArr)
-            case Some(
-                  ParseHeadersIncomplete(
-                    _,
-                    accHeaders,
-                    idx,
-                    state,
-                    name,
-                    start,
-                    chunked,
-                    contentLength)) =>
-              headersInSection(nextArr, idx, state, accHeaders, chunked, contentLength, name, start)
-          }
-          result match {
-            case ParseHeadersCompleted(headers, rest, chunked, length) =>
-              F.pure((headers, chunked, length, rest))
-            case p @ ParseHeadersError(_) => F.raiseError(p)
-            case p @ ParseHeadersIncomplete(_, _, _, _, _, _, _, _) =>
-              if (nextArr.size <= maxHeaderLength)
-                parseHeaders(Array.emptyByteArray, read, maxHeaderLength, p.some)
-              else
-                F.raiseError(
-                  ParseHeadersError(
-                    new Throwable(
-                      s"Parse Headers Exceeded Max Content-Length current size: ${nextArr.size}, only allow ${maxHeaderLength}")
-                  )
-                )
-          }
-        case None =>
-          F.raiseError(
-            ParseHeadersError(new Throwable("Reached End of Stream Looking for Headers")))
+  object MessageP {
+
+    val emptyStreamError = new Throwable("Cannot Parse Empty Stream")
+
+    private val cr: Byte = '\r'.toByte
+    private val lf: Byte = '\n'.toByte
+    private val DoubleCrlf: Seq[Byte] = Seq(cr, lf, cr, lf)
+
+    def parseMessage[F[_]](buffer: Array[Byte], read: F[Option[Chunk[Byte]]], maxHeaderSize: Int)(
```

Comment:
I think this can be made better in a followup. We don't need to linearly scan the entire array each time we get a new chunk, we can start from 3 bytes before the existing buffer.